### ISSUE TITLE
Manually loaded anonymous modules will throw in the next loader request 

### ISSUE THREAD

**Issue by jbalsas** [liferay/liferay-amd-loader#65]
Right now, if an anonymous module is loaded manually using `<script>`, the loader will try to name it on the next script load and will throw an exception.

**Steps to reproduce:**

Update the [demo](https://github.com/liferay/liferay-amd-loader/blob/master/src/demo/index.html#L14) adding:

```html
<script src="//localhost:8080/demo/modules3/a.js"></script>
```

It will throw the `Multiple anonymous modules cannot be served...` error on the first loader `scriptload`.

I'm not sure if this is an issue or something we can do at all, but the logic seems a bit off, since it will always throw. It doesn't seem to affect anything else, so maybe we should just signal it with a `console.warn` instead of throwing an error.

What do you think?

**Comment by ipeychev** [comment]
Hey Chema,

This is expected and it is not an error. Please read [this](https://github.com/liferay/liferay-amd-loader/issues/60#issuecomment-213447520) and [this](https://github.com/liferay/liferay-amd-loader/issues/60#issuecomment-213764069) comment.

Thanks,

**Comment by jbalsas** [comment]
I see, so this will behave exactly as RequireJS does?

**Comment by ipeychev** [comment]
Maybe the error message will be different, but there is no way to achieve this.
The whole conception about anonymous modules is just stupid :)

**Comment by jbalsas** [comment]
Hahaha... yeah, I know, I was just trying to see if a less severe action could be taken, like `console.warn`.

For reference, I was updating Liferay to use the new `v1.5.0` version and in there, we are manually loading `lodash`, which [defines itself as anonymous](https://github.com/lodash/lodash/blob/master/dist/lodash.js#L16128-L16135), so the loader was throwing on subsequent requests.

I've patched our lodash version so it never tries to register for now. I'll try to make the build name the define later on.

Thanks!

**Comment by ipeychev** [comment]
I just checked to be sure. Using the following code:
```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title></title>
  <link rel="stylesheet" href="">

  <script type="text/javascript" src="require.js"></script>
  <script type="text/javascript" src="lodash.js"></script>
</head>
<body>
  
</body>
</html>
```

The error is:
```
require.js:168 Uncaught Error: Mismatched anonymous define() module: function () {
      return _;
    }
http://requirejs.org/docs/errors.html#mismatchmakeError @ require.js:168intakeDefines @ require.js:1251(anonymous function) @ require.js:1449setTimeout (async)(anonymous function) @ require.js:1812localRequire @ require.js:1446requirejs @ require.js:1794(anonymous function) @ require.js:1833(anonymous function) @ require.js:2142
```

Which just proves that anonymous AMD modules were invented by the Devil or at least he has finger in the pie :)

**Comment by jbalsas** [comment]
👍

**Comment by jbalsas** [comment]
Thanks a lot for all the help and attention you put into this!

**Comment by yuchi** [comment]
Actually I find anonymous modules to be the right way to do AMD. Better have no name than a name that could mislead.

**Comment by yuchi** [comment]
(Mislead the loader of course)

**Comment by ipeychev** [comment]
I'm not sure what you mean with "Better have no name than a name that could mislead the loader". What can mislead the loader (and the developer too) are exactly the anonymous modules. Even this issue is a very good example for such a confusion!

Anonymous modules have many disadvantages:
1) They cannot be loaded using a combo service so you end up with:
- hundreds of requests or
- sloppy workarounds
2) You cannot have deferred define - to wrap the define in `setTimeout` or to define the module after some Promise fulfills.
3) You cannot load them explicitly (manually).

But everybody is free to choose whatever makes him/her happy, we live in a free society.

**Comment by yuchi** [comment]
IMHO named modules, **when authored that way**, put a burden on the author that can become a burden on the consumer if there are naming misalignments, and make versioned require a nightmare.

**Comment by jbalsas** [comment]
Hey @ipeychev, I know how much you hate anonymous modules, but I'm reopening this one, not so much because of the error, but because I found out that we may name modules wrong with our current logic 😢 

If the next require request has no dependencies, instead of throwing, we'll assigning that new module the anonymous module implementation, and lose the other one. You can assert this with the following `demo.html`

```html
<!DOCTYPE html>
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title>Test Alloy Loader</title>
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">

    </head>
    <body>
        <h1>Alloy Loader like test page</h1>

        <script src="//localhost:8888/demo/config/config.js"></script>
        <script src="//localhost:8888/loader-debug.js"></script>
        <script src="//localhost:8888/demo/modules3/a.js"></script>

        <script>
            __CONFIG__.maps['mappedModule'] = 'mapped-module';
            __CONFIG__.paths['mapped-module'] = '/modules2/mapped-module';

            __CONFIG__.paths['one'] = '/modules2/one';
            __CONFIG__.paths['two'] = '/modules2/two';
            __CONFIG__.paths['three'] = '/modules2/three';
            __CONFIG__.paths['impl_as_object'] = '/modules2/impl_as_object';

            require(['one'], function (one) {
                console.log(one.toString());
                one('Hello world!');
            }, function(error) {
                console.error(error);
            });

            require(['one'], function (one) {
                console.log(one.toString());
                one('Hello world!');
            }, function(error) {
                console.error(error);
            });
        </script>
    </body>
</html>
```

I'm looking into this now, but any hints would be welcome :)

**Comment by jbalsas** [comment]
I was wondering if we could just ignore manually loaded anonymous modules. That will mean that if the module is configured and then requested again, a second request will happen, but the names and implementations will be assigned properly.

**Comment by jbalsas** [comment]
We could also use [currentScript](https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript) to name them when available and just ignore the rest of them...

**Comment by jbalsas** [comment]
Hey @ipeychev, I've sent #66. It is a proof of concept that should at least mitigate the issue while keeping the original design.

**Comment by ipeychev** [comment]
Hey Chema,

I think we cannot use [currentScript](https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript), because is not supported in IE, is it?
I committed a possible [fix](https://github.com/liferay/liferay-amd-loader/commit/6183b09bec673bf8f46dad952395ff7b1c861a6c), could you please tell me if it works for you too?

About the anonymous modules - yeah, I hate them with passion. The inventor of anonymous modules should receive a condom as a gift, to not reproduce. 

Thanks,

**Comment by yuchi** [comment]
Savage!

**Comment by jbalsas** [comment]
Hey @ipeychev, I guess that fixes the issue with the missmatched module, but I got the feeling we might do a bit more for anonymous modules with `currentScript`. We can actually name them without deferring to a next `scriptLoad` event that may or may not happen...

It's true that is not supported in IE, that's why I said it was a proof of concept, and that it will improve only when it was available. I think I've seen some approaches that take `get the last parsed script` as the closest version possible for IE, so we could fall back to that...

I'm off the office today, so I'll take a look at this tomorrow. 

In any case, I think you could go ahead and release a version, we can add another patch later on if we decide to move forward with this.

Thanks!

**Comment by ipeychev** [comment]
Hey Chema,

We can use `scriptElement` but it won't be the same and you cannot have a real polyfil. This means we will have two different behaviors depending on if it is supported or not and it is better to not do it.

Let's release a new version once you confirm it works for you. I can add a tag if pointing directly to master is not desirable.

Thanks,

**Comment by jbalsas** [comment]
Hey Iliyan,

Yeah, I did consider for a while having a slightly different behaviour, but you're right, that's just messy, let's keep things consistent.

I just verified and the fix works as expected.

There's just one more thing. It was brought up by some guys in our Support team that this level of logging may be too aggressive for some customers and existing setups. I'm fine with the current behaviour, as we discussed, but I was thinking about a way to allow users of the loader to opt-out this errors. Something like:

```javascript
__CONFIG__.throwsOnErrors = true
```

or more fine grained

```javascript
// option 1
__CONFIG__.throwsOn = ['MISMATCHED_ANONYMOUS_MODULE`];

// option 2
__CONFIG__.throwsOn = {
    mismatchedAnonymousModules: true
}
```

what do you think?

**Comment by ipeychev** [comment]
Hey Chema,

About the logging, no problem - what should happen when there is a mismatched anonymous module is not defined in AMD specification, so we can do whatever we want.
RequireJS throws and I think it cannot be changed, but we can do something like:
```__CONFIG__.reportMismatchedAnonymousModules = "exception" or "log"```

Maybe a bit verbose, but at least clear enough.

```__CONFIG__.throwsOn``` is not bad at all, bur I think there are no "other errors" :) The loader is supposed to report the errors by calling the error callback, passed in `require`. Mismatch anonymous module error will however appear in `define` so we can do a config property especially for them.

How does this sound for you?

Thanks,

**Comment by jbalsas** [comment]
Hey Iliyan, `__CONFIG__.reportMismatchedAnonymousModules = "exception|log"` sounds perfect. Will do that tomorrow and send a PR so we can release `1.5.1`.

Thanks a lot! ☺️

**Comment by ipeychev** [comment]
Okay. Please keep in mind that currently all `console` statements are removed in production and the clients which will replace the exceptions with logging, will not see anything if they are using the minified version.

**Comment by jbalsas** [comment]
Done! Found a way to cheat `stripDebug` 😉 -> https://github.com/liferay/liferay-amd-loader/pull/67

**Comment by ipeychev** [comment]
Fixed in [v1.5.1](https://github.com/liferay/liferay-amd-loader/releases/tag/v1.5.1)
